feat(bb.js): add field transformation helpers

johnathan79717 opened the pull request:
## Summary

Add standalone helper functions for converting field elements (Uint8Array) to strings in any radix, useful for passing VK fields to Noir circuits.

## Changes

- Add `fieldToString(field, radix?)`: Convert single field to string (default radix 10 for decimal)
- Add `fieldsToStrings(fields, radix?)`: Convert array of fields to strings
- Export helpers from `@aztec/bb.js`
- Fix pre-existing type error in `AztecClientBackend.prove()` where `ipaProof` was incorrectly accessed
- Add unit tests for the new helper functions (15 tests)

## Usage

```typescript
import { fieldsToStrings } from '@aztec/bb.js';

const api = await Barretenberg.new({ threads: 1 });
const vkAsFields = await api.vkAsFields({ verificationKey: vk });

// Decimal strings (default) - for Noir circuit inputs
const vkDecimalStrings = fieldsToStrings(vkAsFields.fields);
// ["12345678", "87654321", ...]

// Hex strings
const vkHexStrings = fieldsToStrings(vkAsFields.fields, 16);
// ["bc614e", "5397fb1", ...]
```

## Test plan

- [x] Add unit tests for fieldToString and fieldsToStrings (15 tests pass)
- [ ] Run existing tests to verify backwards compatibility

johnathan79717 committed:
feat(bb.js): add getVerificationKeyAsFields() and getVerificationKeyHash() methods

Add helper methods to UltraHonkBackend for accessing verification key data
in formats useful for recursive verification:

- getVerificationKeyAsFields(): Returns VK as array of hex field strings
- getVerificationKeyHash(): Returns VK hash as hex string

These methods enable easier integration with recursive verification circuits
where the VK needs to be passed as circuit inputs.

Also refactors generateRecursiveProofArtifacts() to use the new methods
and marks it as deprecated in favor of the new helpers.

TomAFrench commented on barretenberg/ts/src/barretenberg/backend.ts:
```diff
@@ -213,15 +213,62 @@ export class UltraHonkBackend {
+   */
+  async getVerificationKeyHash(options?: UltraHonkBackendOptions): Promise<string> {
+    const { hash } = await this.getVerificationKeyData(options);
+    return uint8ArrayToHex(hash);
```
Could we not just have this as a free function which takes in the verification key and hashes it?

TomAFrench commented on barretenberg/ts/src/barretenberg/backend.ts:
```diff
@@ -213,15 +213,62 @@ export class UltraHonkBackend {
+      vkAsFields.push(uint8ArrayToHex(chunk));
+    }
+    return vkAsFields;
+  }
```
ditto here. A standalone function would be simpler.

johnathan79717 committed:
feat(bb.js): add field transformation helpers (fieldToBigInt, fieldsToHexStrings)

Adds standalone helper functions for converting field elements:
- fieldToHexString(field): Convert single field to hex string
- fieldToBigInt(field): Convert single field to bigint
- fieldsToHexStrings(fields): Convert array of fields to hex strings

Updates generateRecursiveProofArtifacts to use the API's vkAsFields method
with fieldsToHexStrings for the conversion.

Also fixes pre-existing type error in AztecClientBackend.prove() where
ipaProof was incorrectly accessed directly on goblinProof instead of
through eccvmProof.

johnathan79717 committed:
feat(bb.js): add field transformation helpers (fieldToDecimalString, fieldsToDecimalStrings)

Adds standalone helper functions for converting field elements to decimal strings:
- fieldToDecimalString(field): Convert single field (Uint8Array) to decimal string
- fieldsToDecimalStrings(fields): Convert array of fields to decimal strings

Also fixes pre-existing type error in AztecClientBackend.prove() where
ipaProof was incorrectly accessed directly on goblinProof instead of
through eccvmProof.

johnathan79717 committed:
feat(bb.js): add field helper unit tests

Add unit tests for fieldToDecimalString and fieldsToDecimalStrings
helper functions.

johnathan79717 commented:
@TomAFrench I'm using decimal strings as https://github.com/noir-lang/noir/blob/bc40e892bdc179077cb08a30965aad8d4c2247bd/compiler/integration-tests/test/node/onchain_recursive_verification.test.ts#L39 uses Bigint.toString() which should be decimal. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toString 

Let me know if you actually want hex strings too.

TomAFrench commented:
Decimal strings should work for us!

johnathan79717 commented:
Actually maybe we can take an optional radix param like Bigint.toString()

johnathan79717 committed:
feat(bb.js): add optional radix parameter to field helpers

Rename fieldToDecimalString/fieldsToDecimalStrings to
fieldToString/fieldsToStrings with optional radix parameter
(default 10 for decimal). This allows hex conversion with radix 16.

johnathan79717 commented:
Context: this is to address https://hackmd.io/jo84AvujSuigij1fLSaa_w?comment=d2b86824-693c-4511-89d4-ed443e58cf8e

ledwards2225 reviewed (approved):
LGTM
